### PR TITLE
lxd_container: support lxd instance types

### PR DESCRIPTION
**Pull request by rchicoli**
##### SUMMARY

This MR adds support to lxd_container module to deploy either containers or virtual-machines. For that it was necessary to create two new parameters `api_endpoint` and `type`:

* `api_endpoint`: is required in order to be backwards compatible with old LXD servers, which does not support the new API endpoint `/1.0/virtual-machines` or `/1.0/instances` (implemented in [lxd 3.19](https://discuss.linuxcontainers.org/t/lxd-3-19-has-been-released/6529))
* `type`: there is two different types of instances, “container” or “virtual-machine”

It is related to:
* https://github.com/ansible-collections/community.general/pull/3519

It is a different approach of an open MR:
* https://github.com/ansible-collections/community.general/pull/3480

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

lxd_container

##### ADDITIONAL INFORMATION

The old functionality has not changed. Additionally I've tested the changes only with the latest LXD version 4.19.
* it is possible to create virtual-machine or container
* if an instance already exists and someone tries to update its "type" from `container` to `virtual-machine` or vice-versa, it fails with a warning, otherwise I was only getting changed as status, which is not correct.

### CODE REVIEW

**Comment by conloos** [comment]
Hi @rchicoli,

there are several people who are currently working on providing the functionality of the 4.x lxd-api version.
See: [lxd_virtual_machine - new module #3480](https://github.com/ansible-collections/community.general/pull/3480).
@Felix I prefer to have only one plugin. My suggestion is to rename lxd_container to lxd, as far as possible and e.g. add a symbolic link for backward compatibility.

It is to late so i didn't look deep on the changes.


Thanks Frank

**Comment by rchicoli** [comment]
Hi @conloos 

I've seen the open MR 3480, but like I mentioned I've took a different approach. If the machine "type" is given, then it uses `/containers` or `/virtual-machines`.
It is also possible to change the default API endpoint to `/1.0/instances`, at the end it will work in the same way, but I guess it would be easier to understand the API calls, if someone is following the LXD documentation

FYI: this new API exists since LXD v3.19.

I wouldn't rename or symlink to lxd, because is too generic. For example, if we would create a different module for managing LXC Volumes, then I would expect another module named `lxc_volume`, just like [lxd_profile](https://docs.ansible.com/ansible/latest/collections/community/general/lxd_profile_module.html). For that reason my suggestion are:

1. to keep the module name as lxd_container, as the LXC CLI stands for Linux Containers
2. to rename to lxd_instances (similar to the new API endpoint), symlink to lxd_container for backward compatibility and add the deprecation warning

What do think?

Have a good day

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [inline plugins/modules/cloud/lxd/lxd_container.py:33]: I've seen in the documentation that there are multiple lines that reference to container and not an instance. I can update this tonight, if this MR is welcome

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [inline plugins/modules/cloud/lxd/lxd_container.py:16]: add `.` to the end of the line

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [inline plugins/modules/cloud/lxd/lxd_container.py:119]: add `.` to the end of the line

**Comment by conloos** [comment]
Hi @rchicoli,

I looked at your changes and was wondering if "api_endpoint" is needed at all.
By checking whether the branch "/1.0/instances" exists, it can, from my point of view, be configured automatically so that the user does not have to do a configuration by hand and we minimize the configuration options.

I agree that all documentation have to be updated, like it is done in #3519.

Greetings Frank

**Comment by felixfontein** [comment]
I think something like `api_endpoint` should not really be exposed to users, it should be selected somehow by the module (based on auto-detection and/or user input).

**Comment by rchicoli** [comment]
Ok. the changes should always be backwards compatible, imho to give the user the opportunity to switch to a new endpoint would have been the better choice. Let me try to explain.

I see only one problem. If the user deploys a container and for whatever reason, he changes its "type" to virtual-machine and set the "state" to absent, then the container won't be deleted and Ansible will report that there is nothing to be changed.
The reason is, if you query `/1.0/instances`, LXD responses everything. Now if the type has changed, then the internal API endpoints will be switched to `/1.0/virtual-machines`, LXD says, nothing found. Ansible is happy, because he thinks the instance does not exist.
I guess you get my point. If the API would be /instances LXD is clever enough to remove the running instance, even though the type wouldn't match

Anyways, I will revert that..

- remove the /1.0/instances api
- and reuse only /1.0/containers or /1.0/virtual-machines based on the type provided

**Comment by rchicoli** [comment]
I pushed the changes, but it is running late now. So I would probably double check tomorrow.
After running some small tests locally, everything seems to be working as expected.

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:113]:
```suggestion
        default: container
        choices:
          - container
          - virtual-machine
```

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:391]: Are you sure that this parameter belongs in this list?

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:425]:
```suggestion
```
With `choices` Ansible does this validation for you.

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:752]:
```suggestion
                default='container',
                choices=['container', 'virtual-machine'],
```

File: changelogs/fragments/3661-lxd_container-add-vm-support.yml
Comment (felixfontein) [inline changelogs/fragments/3661-lxd_container-add-vm-support.yml:2]:
```suggestion
  - lxd_container - adds ``type`` option which also allows to operate on virtual machines and not just containers (https://github.com/ansible-collections/community.general/pull/3661).
```

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:114]:
```suggestion
        type: str
        version_added: 4.1.0
```

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [reply plugins/modules/cloud/lxd/lxd_container.py]: perfect, less code with the same functionality.

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [reply plugins/modules/cloud/lxd/lxd_container.py]: you're right, now as long as we are changing the API based on the resource type, it is not required to add it extra to the request. It was for the instances endpoint. Good catch.

**Comment by rchicoli** [comment]
hey @felixfontein  thanks for the review. 
The changes are already done. Please let me know if you find something else, otherwise I would rebase and force push everything.

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (rchicoli) [inline plugins/modules/cloud/lxd/lxd_container.py:664]: todo: check if this is still valid. If I see it right, it would only work with the /instances api

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:344]:
```suggestion
  description: Mapping from the network device name to a list of IPv4 addresses in the instance.
```

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:349]:
```suggestion
  description: The old state of the instance.
```

File: plugins/modules/cloud/lxd/lxd_container.py
Comment (felixfontein) [inline plugins/modules/cloud/lxd/lxd_container.py:422]:
```suggestion
        if self.type == 'container':
```

**Review by felixfontein — COMMENTED** [review]
Looks good to me. Just some nit comments. @conloos what do you think?

**Comment by russoz** [comment]
Hi, sorry for not participating in the discussion earlier but, isn't a bit misleading to have a module named `lxd_container` that can manage something that is *not* a container? I see all the `replace("container", "instance")` throughout the file, but the module name remains the same. Wouldn't it be more consistent to copy this PR's code to a new `lxd_instance` and deprecate `lxd_container`? Just my $0.02.

Otherwise, disregarding that terminology conundrum, it LGTM.

**Comment by felixfontein** [comment]
> there are several people who are currently working on providing the functionality of the 4.x lxd-api version. See: [lxd_virtual_machine - new module #3480](https://github.com/ansible-collections/community.general/pull/3480). @felix I prefer to have only one plugin. My suggestion is to rename lxd_container to lxd, as far as possible and e.g. add a symbolic link for backward compatibility.
> 
> It is to late so i didn't look deep on the changes.

@russoz at the beginning of this PR @conloos wrote the above. I think renaming it is a good idea (I don't care whether `lxd` or `lxd_instance`), instead of having suddenly two mostly identical modules to maintain.

**Comment by conloos** [comment]
Hello @all,

from my point of view everything looks fine and the changes don't break anything.

In the next step, the module should be completely reviewed/rewritten.
I checked the swagger documentation and the endpoints "containers" and "virtual-machines" are not listed. So I think there are only for backward compatibility. So in the next release, the module should be check the exported endpoints and use "containers" for compatibility and "instances" if there is support for.

I can make that, but if @rchicoli is interested I'm fine with that.

The second thing is the name of the module. I vote for "renaming to: lxd_instance and set a deprecate message by using lxd_container" as @rchicoli it has already suggested. Maybe a symlink and check which name was used to call up the module is possible? 

Thanks Frank

**Comment by felixfontein** [comment]
> from my point of view everything looks fine and the changes don't break anything.

In that case I will merge once the final nits are fixed.

> The second thing is the name of the module. I vote for "renaming to: lxd_instance and set a deprecate message by using lxd_container" as @rchicoli it has already suggested. Maybe a symlink and check which name was used to call up the module is possible?

The symlink will mainly be needed for Ansible 2.9, for ansible-base 2.10 and later an entry in meta/runtime.yml will be needed. I can help with the rename; if someone starts a PR for the rename which renames the module (and does all the documentation updates etc. that are necessary) and keeps the checkbox for maintainer pushes enabled, I can push the remaining technical parts to it.

**Comment by rchicoli** [comment]
> In the next step, the module should be completely reviewed/rewritten.

I agree.

> I checked the swagger documentation and the endpoints "containers" and "virtual-machines" are not listed. So I think there are only for backward compatibility. So in the next release, the module should be check the exported endpoints and use "containers" for compatibility and "instances" if there is support for.

Exactly. I have the feeling they didn't want to duplicate the swagger api to add the "containers" and "virtual-machines" endpoints. So you are wondering here is the API implementation:
https://github.com/lxc/lxd/blob/e6523c33b1c6987f794623a5abbc857b8a32a729/lxd/instances.go#L41-L45

> I can make that, but if @rchicoli is interested I'm fine with that.

you can take over this time. At the moment I am kinda busy with other projects.

I want to be a bit precautious before squashing all commits, since it is my first contribution.
So would it be fine to rebase everything now?

If I understood it right, further changes will be done as a nextt step in a separated MR.

I am glad to contribute to the code. Thanks everyone for the support.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
Merging since there's an implicit 'shipit' by @conloos in https://github.com/ansible-collections/community.general/pull/3661#issuecomment-968075362 (sorry, missed that).

**Comment by felixfontein** [comment]
@rchicoli thanks for your contribution!
@conloos @russoz thanks for reviewing!

**Pull request merged**
